### PR TITLE
Add android nullness annotations

### LINKED ISSUE #96: android annotation support?
As I know: 
Findbugs has added support for these annotations:
@com.google.common.base.Nullable
@org.jetbrains.annotations.Nullable
@org.jetbrains.annotations.NotNull
@org.eclipse.jdt.annotation.Nullable
@org.eclipse.jdt.annotation.NonNull
@org.eclipse.jdt.annotation.NonNullByDefault
+
JSR305

but  android-related:
android.support.annotation.NonNull;
android.support.annotation.Nullable;

is not supported-yet?

If I want to add support for android annotation, how can I do?

### PR DESCRIPTION
**Pull request by kzaikin**
The pull request is built from presumably small commits

Commit 1 adds tests on existing NullnessAnnotations inside spotbugs gradle project
Commit 2 adds android annotations to NullnessAnnotations along with a small refactoring (we already have tests so it's safe)
Commit 3 adds android annotations to TypeQualifierResolver and a few integration tests using SpotBugsRule approach

Fix #96

### CODE REVIEW

File: spotbugs-tests/build.gradle [hunk 1]
```diff
@@ -20,14 +10,15 @@ dependencies {
...
 
+  compile project(':spotbugs')
   // These annotations are repackaged to spotbugs.jar, to keep backward compatibility for Ant task.
   // If they're not repackaged, Ant task will report 'java.lang.ClassNotFoundException: edu.umd.cs.findbugs.annotations.CleanupObligation'
   compile project(':spotbugs-annotations')
+  compile project(':spotbugsTestCases')
+  compile project(':test-harness')
 
   compileOnly 'com.apple:AppleJavaExtensions:1.4'
-  compile 'junit:junit:4.12'
```
Comment (kzaikin) [inline spotbugs-tests/build.gradle]: this dependency will come from test-harness

File: spotbugs-tests/build.gradle [hunk 2]
```diff
@@ -36,16 +27,6 @@ task sourcesJar(type: Jar) {
   from sourceSets.main.allSource
 }
 
-test {
-// I have no idea about gradle test execution and how to properly give it the
-// required dependencies
-// This seem to work with most tests except the new one based on AnalysisRunner
-// TODO fix AnalysisRunner or the test classpath
-    classpath = project.sourceSets.main.runtimeClasspath \
-        + project(':spotbugs').sourceSets.main.runtimeClasspath \
```
Comment (kzaikin) [inline spotbugs-tests/build.gradle]: the true way of referencing classes is adding dependency on a project

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/DetectorsTest.java [hunk 3]
```diff
@@ -214,7 +214,6 @@ private void setUpEngine(String... analyzeMe) throws IOException {
             }
         }
 
-        project.addAuxClasspathEntry("lib/junit.jar");
```
Comment (kzaikin) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/DetectorsTest.java]: this is not necessary as soon as junit is in classpath via project dependencies

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/DetectorsTest.java [hunk 4]
```diff
@@ -236,7 +235,9 @@ private void setUpEngine(String... analyzeMe) throws IOException {
             String[] cpParts = classpath.split(File.pathSeparator);
             for (String cpStr : cpParts) {
                 File file = new File(cpStr);
-                project.addAuxClasspathEntry(file.getCanonicalPath());
```
Comment (kzaikin) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/DetectorsTest.java]: classpath might contain empty directories witch make project unhappy

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,10 @@ This is the changelog for SpotBugs. This follows [Keep a Changelog v0.3](http://
 
 ## Unreleased (2017/??/??)
 
+### Added
+
+* Make TypeQualifierResolver recognize android.support.annotation.NonNull and Nullable ([#182](https://github.com/spotbugs/spotbugs/pull/182))
```
Comment (kzaikin) [inline CHANGELOG.md]: added it

**Comment by kzaikin** [comment]
Should anything else be done to have this merged?

File: spotbugsTestCases/src/fakeAnnotations/android/support/annotation/NonNull.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Denotes that a parameter, field or method return value can never be null.
+ * <p>
+ * This is a marker annotation and it has no specific attributes.
+ */
+@Documented
+@Retention(CLASS)
+@Target({METHOD, PARAMETER, FIELD, ANNOTATION_TYPE, PACKAGE})
+public @interface NonNull {
+}
```
Comment (KengoTODA) [inline spotbugsTestCases/src/fakeAnnotations/android/support/annotation/NonNull.java]: Confirmed that this code has same Retention policy with https://github.com/android/platform_frameworks_support/blob/master/annotations/src/android/support/annotation/NonNull.java

**Comment by scheffield** [comment]
This change is really great. Unfortunately the existing gradle plugin in version 1.1 still points to RC2 which makes this change unavailable for Android builds, right?

**Comment by KengoTODA** [comment]
@scheffield You can use `spotbugs.toolVersion` to specify the version of SpotBugs, please have a try:

http://spotbugs.readthedocs.io/en/latest/migration.html#findbugs-gradle-plugin

**Pull request merged**